OpenRA.Game/Network/LocalizedMessage.cs

```diff
@@ -59,7 +59,7 @@ public class LocalizedMessage
 		[FieldLoader.LoadUsing(nameof(LoadArguments))]
 		public readonly FluentArgument[] Arguments = Array.Empty<FluentArgument>();
 
-		public string TranslatedText { get; }
+		public Dictionary<string, object> ArgumentsDictionary { get; }
```

Comment:
Can't this be done by changing the type of the above from `FluentArgument[]` to `Dictionary<string, object>` and doing the things we currently do in the constructor in `LoadArguments`? Then this here can be removed and we can directly access `.Arguments` which has what we want.

Reply:
No, that won't work. In serializing the `Dictionary`, we lose the object type.

Reply:
I don't understand, if having a fluent type instead of an object is a problem, why not change the loading part to convert that to an object or load it as an object instead?

Reply:
```diff
diff --git a/OpenRA.Game/Network/LocalizedMessage.cs b/OpenRA.Game/Network/LocalizedMessage.cs
index fc48febf0a..b56f3b181c 100644
--- a/OpenRA.Game/Network/LocalizedMessage.cs
+++ b/OpenRA.Game/Network/LocalizedMessage.cs
@@ -57,43 +57,36 @@ public class LocalizedMessage
                public readonly string Key = string.Empty;
 
                [FieldLoader.LoadUsing(nameof(LoadArguments))]
-               public readonly FluentArgument[] Arguments = Array.Empty<FluentArgument>();
-
-               public Dictionary<string, object> ArgumentsDictionary { get; }
+               public readonly Dictionary<string, object> Arguments;
 
                static object LoadArguments(MiniYaml yaml)
                {
-                       var arguments = new List<FluentArgument>();
+                       var arguments = new Dictionary<string, object>();
                        var argumentsNode = yaml.Nodes.FirstOrDefault(n => n.Key == "Arguments");
                        if (argumentsNode != null)
                        {
-                               var regex = new Regex(@"Argument@\d+");
-                               foreach (var argument in argumentsNode.Value.Nodes)
-                                       if (regex.IsMatch(argument.Key))
-                                               arguments.Add(FieldLoader.Load<FluentArgument>(argument.Value));
+                               foreach (var argumentNode in argumentsNode.Value.Nodes)
+                               {
+                                       var argument = FieldLoader.Load<FluentArgument>(argumentNode.Value);
+                                       if (argument.Type == FluentArgument.FluentArgumentType.Number)
+                                       {
+                                               if (!double.TryParse(argument.Value, out var number))
+                                                       Log.Write("debug", $"Failed to parse {argument.Value}");
+
+                                               arguments.Add(argument.Key, number);
+                                       }
+                                       else
+                                               arguments.Add(argument.Key, argument.Value);
+                               }
                        }
 
-                       return arguments.ToArray();
+                       return arguments;
                }
 
                public LocalizedMessage(MiniYaml yaml)
                {
                        // Let the FieldLoader do the dirty work of loading the public fields.
                        FieldLoader.Load(this, yaml);
-
-                       ArgumentsDictionary = new Dictionary<string, object>();
-                       foreach (var argument in Arguments)
-                       {
-                               if (argument.Type == FluentArgument.FluentArgumentType.Number)
-                               {
-                                       if (!double.TryParse(argument.Value, out var number))
-                                               Log.Write("debug", $"Failed to parse {argument.Value}");
-
-                                       ArgumentsDictionary.Add(argument.Key, number);
-                               }
-                               else
-                                       ArgumentsDictionary.Add(argument.Key, argument.Value);
-                       }
                }
 
                public static string Serialize(string key, Dictionary<string, object> arguments = null)
```

Reply:
This commit is doing too many distinct things, and should be split into 3:

1. Add missing `[TranslationReference]`s.
2. `TranslationProvider` and `LocalizedMessage` changes.
3. Extracting hardcoded strings to fluent.

Reply:
first part of 2nd point was done in #20804. As for 1st, they aren't really missing before translations are extracted.